runtime/lua/vim/lsp/document_color.lua

```diff
@@ -0,0 +1,386 @@
+local api = vim.api
+local lsp = vim.lsp
+local util = lsp.util
+local ms = lsp.protocol.Methods
+
+local document_color_ns = api.nvim_create_namespace('nvim.lsp.document_color')
+local document_color_augroup = api.nvim_create_augroup('nvim.lsp.document_color', {})
+
+local M = {}
+
+--- @alias vim.lsp.document_color.HighlightInfo { hex_code: string, hl_group?: string, range: Range4 }
+
+--- @class (private) vim.lsp.document_color.GlobalState
+--- @field enabled boolean
+local global_state = { enabled = false }
+
+--- @class (private) vim.lsp.document_color.BufState : vim.lsp.document_color.GlobalState
+--- @field buf_version? integer Buffer version for which the color ranges correspond to.
+--- @field applied_version? integer Last buffer version for which we applied color ranges.
+--- @field hl_infos? table<integer, vim.lsp.document_color.HighlightInfo[]> client_id -> processed color highlights
+
+--- @type table<integer, vim.lsp.document_color.BufState>
+local bufstates = vim.defaulttable(function(_)
+  return setmetatable({}, {
+    __index = global_state,
+    __newindex = function(state, key, value)
+      if global_state[key] == value then
+        rawset(state, key, nil)
+      else
+        rawset(state, key, value)
+      end
+    end,
+  })
+end)
+
+--- @inlinedoc
+--- @class vim.lsp.document_color.enable.Opts
+---
+--- Highlight style. It can be one of the pre-defined styles, a string to be used as virtual text, or a
+--- function that receives the buffer handle, the range (start line, start col, end line, end col) and
+--- the resolved hex color.
+--- Defaults to 'background'.
+--- @field style? string|fun(bufnr: integer, range: Range4, hex_code: string)
+
+-- Default options.
+--- @type vim.lsp.document_color.enable.Opts
+M._opts = { style = 'background' }
```

Comment:
Is there a reason this is attached to `M`?

Reply:
So that I can access it inside and outside methods: https://github.com/neovim/neovim/blob/d25c633f95382e92942dd103f5be7671bb5e4d37/runtime/lua/vim/lsp/document_color.lua#L358

Reply:
Surely you can still access that if this was made a `local`? I can't see this being accessed outside of this file.